### ISSUE TITLE
Make store hooks be executed in parallel if possible

### ISSUE THREAD

**Issue by matthiasbeyer** [matthiasbeyer/imag#303]
The store currently executes hooks in a non-parallel way.

That's because the _Store_ object itself does not know whether an aspect wants its hooks to be executed in parallel or not and it also does now know whether the aspect contains hooks which want parallel access.

So, the `Aspect` is capable of executing hooks in parallel if possible.
# How to

Well, first of, the `Aspect` has to check how many non-mutable hooks there are. If there are too much mutable hooks and to few non-mutable, it does not make sense to start threads.

For example if there are 10 mutable hooks but only 2 non-mutable, there's no point in executing the two non-mutable ones in parallel, as we do not gain that much performance here. If it is the other way round, it would make sense to spin up two or three threads, though.

What has to be considered: The order _must_ be kept. We cannot say "Hey, lets collect all non-mutable hooks, execute them first and then execute the mutable ones". This is because the _user_ configures which hook runs when. If the user configured it this way, we cannot be sure whether resorting would yield the results the user expects.

So we have to find _chunks_ of hooks in the hooks array, which want to access the entry in an non-mutable way. These hooks can be parallelized. After all entries in such a chunk are executed, we can continue executing the mutating hooks.

Feel free to ask questions how to start implementing this.
# Idea

I guess these steps would be an idea:
1. Count how many non-mutating Hooks there are
2. See whether it makes sense to spin up threads (total number of hooks vs. number of non-mutating hooks)
3. Spin up a threadpool with N workers (config file?)
4. Execute the hooks

To simplify the whole problem, the precalculations and the instantiation of the threadpool could be done right after the registration of the hooks. This would yield _some_ overhead in the registration process, though I think we can pay this price.
# Additional ideas

The following ideas _could_ be implemented in the same PR, but could also be added in seperate PRs later on:

---
- [ ] Would also be nice to have a `treshhold` value in the config which tells the store at how many Hooks an aspect should start to parallelize. For example, you don't want to spin up threads if there are 3 hooks beeing executed, but you want to spin up threads for 15 hooks. A setting could specify a value for this.

---
- [ ] Another Setting for how many threads should be used would be nice as well. For example something like `"<number of hooks> / VALUE"` threads should be used, but `MAX_HOOK_THREADS` should be considered

---
